### DISCUSSION TITLE
What should the AST.comments field contain?

### DISCUSSION THREAD

**Discussion by nene** [prettier/prettier#14158] [Q&A]
I'm writing a prettier plugin and trying to understand how the prettiers comment algorithm works.

The [documentation](https://prettier.io/docs/en/plugins.html#handling-comments-in-a-printer) says that:

> By default, if the AST has a top-level comments property, Prettier assumes that comments stores an array of comment nodes.

OK, but what sort of data should this `comments` field contain? I assume the following is not quite enough:

```js
comments: ["/* comment1 */", "/* comment2 */"]
```

It probably also needs a location info to be able to associate comments with AST nodes. So I tried something like the following (using the same `range` fields that I'm accessing in `locStart` and `locEnd` functions):

```js
comments: [
  { type: "block_comment", text: "/* comment1 */", range: [0, 14] },
  { type: "block_comment", text: "/* comment2 */", range: [32, 46] },
]
```

Doesn't seem to work. I just get an error: `TypeError: Cannot read properties of undefined (reading 'trim')`

The error gets thrown from `ensureAllCommentsPrinted()` function which tries to perform `comment.value.trim()`. Aha... so it expects a `value` field. But when I change my current `text` field to `value`, it still doesn't manage to print the comment, instead ending up with an error: `Comment "/* comment1 */" was not printed. Please report this error!`

I also tried to implement the `printComment()` function. But it never gets called.

**I'm clearly doing something very wrong, but I have no idea what.**

The docs also say:

> Comments are then attached to these nodes **mutating the AST in the process**

This worries me. Especially as I have no idea what kind of mutations are going to be performed.

**Comment by fisker** [comment]
> I have no idea what kind of mutations are going to be performed.

During comment attach,

https://github.com/prettier/prettier/blob/3ef82ae06b41574d9c45ab2506553bcc00afc019/src/main/comments.js#L220-L222

https://github.com/prettier/prettier/blob/3ef82ae06b41574d9c45ab2506553bcc00afc019/src/common/util.js#L407-L412

(There are some other mutations in those files)

After it's printed.

https://github.com/prettier/prettier/blob/3ef82ae06b41574d9c45ab2506553bcc00afc019/src/main/comments.js#L421


You can freeze comment to see what's going to change.

> I also tried to implement the printComment() function. But it never gets called.

It's called automatically when you call `print()`.

https://github.com/prettier/prettier/blob/3ef82ae06b41574d9c45ab2506553bcc00afc019/src/main/ast-to-doc.js#L145

**Comment by nene** [comment]
Thanks for the answer, but I still don't really get it.

From the documentation I understood that prettier will move comments from the root node in AST to the nodes to which they belong to:

> By default, if the AST has a top-level comments property, Prettier assumes that comments stores an array of comment nodes. Prettier will then use the provided `parsers[<plugin>].locStart/locEnd` functions to search for the AST node that each comment "belongs" to. Comments are then attached to these nodes mutating the AST in the process, and the comments property is deleted from the AST root.

This doesn't seem to be happening, and your response also only describes how the comment node itself is mutated, saying nothing about whether something else in the AST also gets changed.

After parsing the following code:

```sql
SELECT /*comment1*/ 1
```

My parser produces the following AST:

```json
{
  "type": "program",
  "statements": [
    {
      "type": "select_stmt",
      "clauses": [
        {
          "type": "select_clause",
          "selectKw": { "type": "keyword", "text": "SELECT", "name": "SELECT", "range": [0, 6] },
          "options": [],
          "columns": {
            "type": "list_expr",
            "items": [
              { "type": "number_literal", "text": "1", "value": 1, "range": [20, 21] }
            ],
            "range": [20, 21],
          },
          "range": [0, 21]
        }
      ],
      "range": [0, 21]
    }
  ],
  "range": [0, 21],
  "comments": [
    { "type": "block_comment", "value": "/*comment1*/", "range": [7, 19] }
  ]
}
```

After prettier has processed it, and calls the `Printer.print` function, the AST looks as follows, with only the comments field having been changed:

```json
{
  "type": "program",
  "statements": [
    {
      "type": "select_stmt",
      "clauses": [
        {
          "type": "select_clause",
          "selectKw": { "type": "keyword", "text": "SELECT", "name": "SELECT", "range": [0, 6] },
          "options": [],
          "columns": {
            "type": "list_expr",
            "items": [
              { "type": "number_literal", "text": "1", "value": 1, "range": [20, 21] }
            ],
            "range": [20, 21],
          },
          "range": [0, 21]
        }
      ],
      "range": [0, 21]
    }
  ],
  "range": [0, 21],
  "comments": [
    { 
      "type": "block_comment",
      "value": "/*comment1*/",
      "range": [7, 19],
      "placement": "remaining",
      "leading": false,
      "trailing": false,
      "printed": false,
      "nodeDescription": "program"
    }
  ]
}
```

Is that's what's supposed to happen or not? It definitely conflicts with what the documentation says (*the comments property is deleted from the AST root.*)

It seems like Prettier associates the comment with the root node (`"nodeDescription": "program"`). This doesn't look correct. The comment should IMHO be either after the `keyword` node or before the `list_expr` node.

You said that `printComment()` is called automatically when you call `print()`. But how exactly should I call it?

- I tried calling `print()` directly with the comment node. That just ended up recursively calling the `Parser.print()` function, not the `Parser.printComment()` function.

- I am calling `print()` with all the non-comment nodes in this AST. This seems to have no effect.

- I can't call `print()` with the root node, which has the `comments` property (that will result in infinite recursion).

**Comment by nene** [comment]
After reading through lots of Prettier source code and running several experiments I finally figured out the culprit:

I had not implemented the `canAttachComment()` function. So Prettier ended up determining that none of the comments can be attached to any node, defaulting to attaching them to the root AST node.

The documentation is not really clear on what methods one should implement to make use of the Prettier automatic comments handling. All the comments-related methods are marked as optional, which makes one think that Prettier will take care of comments automatically out of the box, but it's not really so. One definitely needs to implement `printComment()` and `canAttachComment()`.

If you fail to implement `printComment()` you get a reasonably useful and actionable error message:

- `TypeError: options.printer.printComment is not a function`

But if you fail to implement `canAttachComment()`, the error won't be helpful at all:

- `Comment "/*com1*/" was not printed. Please report this error!`
